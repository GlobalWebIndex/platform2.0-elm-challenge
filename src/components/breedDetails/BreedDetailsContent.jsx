import React from "react";
import PropTypes from "prop-types";
import { Link, useLocation } from "react-router-dom";
import { createUseStyles } from "react-jss";

const useStyles = createUseStyles({
  breedDetailsWrapper: {
    padding: 20,
  },
  breedDetailsHeading: { marginTop: 0 },
  breedDetailsImgGrid: {
    display: "grid",
    gridGap: 10,
    gridTemplateColumns: "repeat(4, 1fr)",
    marginBottom: 30,
    border: "1px solid #000",
  },
  breedDetailsImg: {
    width: "100",
    height: "100",
    objectFit: "cover",
  },
  breedDetailsDescription: {
    lineHeight: "1.5",
  },
});

const BreedDetailsContent = ({ data = [] }) => {
  let classes = useStyles();
  const location = useLocation();
  return (
    <div className={classes.breedDetailsWrapper}>
      <h2 className={classes.breedDetailsHeading}>
        {data[0]?.breeds[0]?.name || ""}
      </h2>
      <div className={classes.breedDetailsImgGrid}>
        {data.map((breedImage) => (
          <Link
            to={`/cat/${breedImage.id}`}
            state={{ backgroundLocation: location }}
            key={breedImage.id}
          >
            <img
              className={classes.breedDetailsImg}
              src={breedImage.url}
              alt={`Breed id: ${breedImage.id}`}
            />
          </Link>
        ))}
      </div>
      <div className={classes.breedDetailsDescription}>
        {data[0]?.breeds[0]?.description}
      </div>
    </div>
  );
};

BreedDetailsContent.propTypes = {
  data: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      url: PropTypes.string,
      breeds: PropTypes.arrayOf(
        PropTypes.shape({
          description: PropTypes.string,
        })
      ),
    })
  ),
};

export default BreedDetailsContent;
