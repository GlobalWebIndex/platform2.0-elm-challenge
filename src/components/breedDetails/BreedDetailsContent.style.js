import { createUseStyles } from "react-jss";

export const useStyles = createUseStyles({
  breedDetailsWrapper: {
    padding: 20,
  },
  breedDetailsHeading: { marginTop: 0 },
  breedDetailsImgGrid: {
    display: "grid",
    gridGap: 10,
    gridTemplateColumns: "repeat(4, 1fr)",
    marginBottom: 30,
    border: "1px solid #000",
  },
  breedDetailsImg: {
    width: "100%",
    height: "100%",
    objectFit: "cover",
  },
  breedDetailsDescription: {
    lineHeight: "1.5",
  },
});
