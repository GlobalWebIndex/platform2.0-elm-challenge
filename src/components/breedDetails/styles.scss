.images {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));

  img {
    width: 100%;
    height: 100%;
    object-fit: cover;
  }
}

@media screen and (min-width: 768px) {
  .images {
    grid-template-columns: repeat(auto-fit, minmax(300px, 1fr));
  }
}
