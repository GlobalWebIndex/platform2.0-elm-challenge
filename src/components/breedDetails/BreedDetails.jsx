import React, { useEffect } from "react";
import { useParams } from "react-router-dom";
import { useAsync, AsyncStatus } from "../../hooks/useAsync";
import { fetchImagesByBreed } from "../../util/api";
import Loading from "../common/Loading";
import Failure from "../common/Failure";
import BreedDetailsContent from "./BreedDetailsContent";

const BreedDetails = () => {
  const { breedId } = useParams();
  const { data: fetchedBreedDetails, status, error, run } = useAsync();

  useEffect(() => {
    const callback = fetchImagesByBreed(breedId);
    run(callback);
  }, [run, breedId]);

  switch (status) {
    case AsyncStatus.IDLE:
    case "pending":
      return <Loading />;

    case "rejected":
      return <Failure error={error} />;

    case AsyncStatus.RESOLVED:
      return <BreedDetailsContent data={fetchedBreedDetails} />;
    default:
      throw new Error("This shouldn't ever happen");
  }
};

export default BreedDetails;
