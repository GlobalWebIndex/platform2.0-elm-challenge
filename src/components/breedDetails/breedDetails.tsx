import React from 'react';
import {Link} from 'react-router-dom';
import type {BreedImages} from 'cat-lover/types/breeds';
import styles from './styles.scss';

type BreedDetailsProps = {
  breedImages: BreedImages;
};

export const BreedDetails = ({breedImages}: BreedDetailsProps): JSX.Element => (
  <div className={styles.images}>
    {breedImages.map(image => (
      <Link key={image.id} to={`/cats/${image.id}`}>
        <img alt="Breed" src={image.url} />
      </Link>
    ))}
  </div>
);
