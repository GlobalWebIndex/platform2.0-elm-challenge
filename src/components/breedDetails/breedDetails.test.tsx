import React from 'react';
import {MemoryRouter} from 'react-router-dom';
import {render} from '@testing-library/react';
import type {BreedImages} from 'cat-lover/types/breeds';
import {BreedDetails} from './breedDetails';

describe('<BreedDetails />', () => {
  it('should render breed images', () => {
    const mockBreedImages: BreedImages = [
      {
        height: 200,
        id: '1',
        url: 'https://example.com/image1.jpg',
        width: 100
      },
      {
        height: 100,
        id: '2',
        url: 'https://example.com/image2.jpg',
        width: 200
      }
    ];

    const {container} = render(
      <MemoryRouter initialEntries={['/breeds/1']}>
        <BreedDetails breedImages={mockBreedImages} />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });
});
