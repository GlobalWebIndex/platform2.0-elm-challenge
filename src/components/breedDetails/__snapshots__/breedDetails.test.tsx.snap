// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BreedDetails /> should render breed images 1`] = `
<div>
  <div
    class="images"
  >
    <a
      href="/cats/1"
    >
      <img
        alt="Breed"
        src="https://example.com/image1.jpg"
      />
    </a>
    <a
      href="/cats/2"
    >
      <img
        alt="Breed"
        src="https://example.com/image2.jpg"
      />
    </a>
  </div>
</div>
`;
